components/sequencer/snowflake/snowflake_sequencer.go

```diff
@@ -0,0 +1,200 @@
+//
+// Copyright 2021 Layotto Authors
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+package snowflake
+
+import (
+	"context"
+	"database/sql"
+	"errors"
+	"sync"
+	"time"
+
+	"mosn.io/pkg/log"
+
+	"mosn.io/layotto/components/pkg/utils"
+	"mosn.io/layotto/components/sequencer"
+)
+
+type SnowFlakeSequencer struct {
+	metadata   utils.SnowflakeMetadata
+	workerId   int64
+	mu         sync.Mutex
+	db         *sql.DB
+	smap       map[string]chan int64
+	biggerThan map[string]int64
+	logger     log.ErrorLogger
+	ctx        context.Context
+	cancel     context.CancelFunc
+}
+
+func NewSnowFlakeSequencer(logger log.ErrorLogger) *SnowFlakeSequencer {
+	return &SnowFlakeSequencer{
+		logger: logger,
+	}
+}
+
+func (s *SnowFlakeSequencer) Init(config sequencer.Configuration) error {
+	var err error
+	s.metadata, err = utils.ParseSnowflakeMetadata(config.Properties)
+	if err != nil {
+		return err
+	}
+	//for unit test
+	s.metadata.MysqlMetadata.Db = s.db
+
+	s.smap = make(map[string]chan int64)
+	s.biggerThan = config.BiggerThan
+	s.ctx, s.cancel = context.WithCancel(context.Background())
+
+	if s.workerId, err = utils.NewMysqlClient(&s.metadata.MysqlMetadata); err != nil {
+		return err
+	}
+	return err
+}
+
+func (s *SnowFlakeSequencer) GetNextId(req *sequencer.GetNextIdRequest) (*sequencer.GetNextIdResponse, error) {
+	s.mu.Lock()
+	var ok bool
+	seed, ok := s.smap[req.Key]
```

Comment:
the efficiency of lock low, conside to use sync.map？

Reply:
got it

Reply:
I need **seed, ok := s.smap[req.Key]** and **if !ok{}** executed together.So I add **s.mu.Lock()** to it. sync.map may be unecessary